File: solutions/java/triangle/1/src/main/java/Triangle.java
enum TriangleType {
    ISOSCELES, EQUILATERAL, SCALENE
}

class Triangle {
    private final TriangleType triangleType;
    
    Triangle(double side1, double side2, double side3) throws TriangleException {
        long s1 = Double.doubleToLongBits(side1);
        long s2 = Double.doubleToLongBits(side2);
        long s3 = Double.doubleToLongBits(side3);
        if (s1 > s2) {
            s1 ^= s2;
            s2 ^= s1;
            s1 ^= s2;
        }
        if (s1 > s3) {
            s1 ^= s3;
            s3 ^= s1;
            s1 ^= s3;
        }
        if (s2 > s3) {
            s2 ^= s3;
            s3 ^= s2;
            s2 ^= s3;
        }
        double d1 = Double.longBitsToDouble(s1);
        double d2 = Double.longBitsToDouble(s2);
        double d3 = Double.longBitsToDouble(s3);
        if (d3 >= d1 + d2) {
            throw new TriangleException();
        }
        if (d1 == d3) {
            triangleType = TriangleType.EQUILATERAL;
        } else if (d2 == d1 || d2 == d3) {
            triangleType = TriangleType.ISOSCELES;
        } else {
            triangleType = TriangleType.SCALENE;
        }
    }

    boolean isEquilateral() {
        return triangleType == TriangleType.EQUILATERAL; 
    }

    boolean isIsosceles() {
        return triangleType == TriangleType.ISOSCELES
            || triangleType == TriangleType.EQUILATERAL;
    }

    boolean isScalene() {
        return triangleType == TriangleType.SCALENE;
    }

}

File: solutions/typescript/triangle/1/triangle.ts
declare global {
  interface Array<T> {
    toSorted(): Array<T>
    toSorted(compareFn: (a: T, b: T) => number): Array<T>
  }
}

export class Triangle {
  private triangleType: 'equ' | 'iso' | 'sca' | 'not' = 'not'
  
  constructor(...sides: number[]) {
    if (sides.length !== 3) {
      return
    }
    const [shortest, medium, longest] = sides.toSorted((a, b) => a - b)
    if (longest >= shortest + medium) {
      return
    }
    if (longest === shortest) {
      this.triangleType = 'equ'
      return
    }
    if (medium === shortest || medium === longest) {
      this.triangleType = 'iso'
      return
    }
    this.triangleType = 'sca'
  }

  get isEquilateral(): boolean {
    return this.triangleType === 'equ'
  }

  get isIsosceles(): boolean {
    return this.triangleType === 'iso' || this.triangleType === 'equ'
  }

  get isScalene(): boolean {
    return this.triangleType === 'sca'
  }
}
